In [1]:
import numpy as np
from scipy import stats

# Варіант 5 — Дніпро
dnipro = np.array([
    82, 79, 74, 67, 61, 59,
    57, 59, 65, 72, 78, 83
])

norm = 70
alpha = 0.05

print("Варіант 5 — Дніпро")
print("Дані:", dnipro)
print("Норма:", norm)

Варіант 5 — Дніпро
Дані: [82 79 74 67 61 59 57 59 65 72 78 83]
Норма: 70


In [2]:
t_stat, p_value = stats.ttest_1samp(
    dnipro,
    popmean=norm
)

print("Середнє:", dnipro.mean())
print("t-статистика:", t_stat)
print("p-value:", p_value)

if p_value < alpha:
    print("Відхиляємо H₀")
else:
    print("Не відхиляємо H₀")

Середнє: 69.66666666666667
t-статистика: -0.12094486171172247
p-value: 0.9059156763414993
Не відхиляємо H₀


H₀: середня вологість Дніпра дорівнює довідковій нормі 70%.
H₁: середня вологість Дніпра відрізняється від 70%.

Отримано t ≈ -0.121 та p ≈ 0.906.

Оскільки p > 0.05, нульову гіпотезу не відхиляємо.

Отже, за цими 12 місячними значеннями немає статистично
значущих підстав вважати, що середня вологість Дніпра
відрізняється від довідкової норми 70%.

In [3]:
n = len(dnipro)

x_mean = dnipro.mean()
s = dnipro.std(ddof=1)

se = s / np.sqrt(n)

t_manual = (x_mean - norm) / se

print("n =", n)
print("Середнє =", x_mean)
print("s =", s)
print("SE =", se)
print("t вручну =", t_manual)
print("t з scipy =", t_stat)

n = 12
Середнє = 69.66666666666667
s = 9.547330263037681
SE = 2.756076848703533
t вручну = -0.12094486171172245
t з scipy = -0.12094486171172247


Ручний розрахунок дав t ≈ -0.1209.

Функція stats.ttest_1samp() також дала t ≈ -0.1209.

Отже, результати збігаються. Це підтверджує, що
ttest_1samp використовує ту саму формулу t-статистики.

In [4]:
odesa = np.array([
    82, 80, 78, 74, 71, 69,
    67, 68, 72, 76, 80, 83
])

# Спочатку критерій Лівена
lev_stat, lev_p = stats.levene(dnipro, odesa)

print("Дніпро std:", dnipro.std(ddof=1))
print("Одеса std:", odesa.std(ddof=1))
print("Levene statistic:", lev_stat)
print("Levene p-value:", lev_p)

Дніпро std: 9.547330263037681
Одеса std: 5.624621199366553
Levene statistic: 6.813202247191015
Levene p-value: 0.015979720548169957


In [5]:
t_ind, p_ind = stats.ttest_ind(
    dnipro,
    odesa,
    equal_var=False
)

print("Середнє Дніпро:", dnipro.mean())
print("Середнє Одеса:", odesa.mean())

print("t:", t_ind)
print("p-value:", p_ind)

Середнє Дніпро: 69.66666666666667
Середнє Одеса: 75.0
t: -1.6672917551057547
p-value: 0.11293913398547563


За критерієм Лівена отримано p ≈ 0.016.

Оскільки p < 0.05, припущення про рівність дисперсій
відхиляємо. Тому для порівняння середніх використано
Welch t-test з equal_var=False.

Середня вологість Дніпра становить приблизно 69.67%,
а Одеси — 74.17%.

Для t-критерію отримано p ≈ 0.113.

Оскільки p > 0.05, нульову гіпотезу про рівність
середніх не відхиляємо.

Отже, за цими даними статистично значущої різниці
середньої вологості Дніпра та Одеси не виявлено.

In [6]:
kyiv = np.array([
    84, 82, 78, 70, 65, 63,
    61, 63, 68, 75, 81, 85
])

lev_stat2, lev_p2 = stats.levene(
    dnipro,
    kyiv
)

print("Дніпро std:", dnipro.std(ddof=1))
print("Київ std:", kyiv.std(ddof=1))

print("Levene statistic:", lev_stat2)
print("Levene p-value:", lev_p2)

Дніпро std: 9.547330263037681
Київ std: 8.9489291724392
Levene statistic: 0.07624064319378997
Levene p-value: 0.7850353624913227


In [7]:
t_ind2, p_ind2 = stats.ttest_ind(
    dnipro,
    kyiv,
    equal_var=True
)

print("Середнє Дніпро:", dnipro.mean())
print("Середнє Київ:", kyiv.mean())

print("t:", t_ind2)
print("p-value:", p_ind2)

Середнє Дніпро: 69.66666666666667
Середнє Київ: 72.91666666666667
t: -0.8603554360569221
p-value: 0.39887242978394566


У Завданні 4 для порівняння обрано Київ.

Критерій Лівена дав p ≈ 0.785.

Оскільки p > 0.05, немає статистично значущих підстав
вважати дисперсії Дніпра та Києва різними.

Тому використано стандартний незалежний t-критерій
з equal_var=True.

Отримано p ≈ 0.399 > 0.05, тому нульову гіпотезу
про рівність середніх не відхиляємо.

Статистично значущої різниці середньої вологості
Дніпра та Києва за цими даними не виявлено.

На відміну від Завдання 3, де критерій Лівена показав
різні дисперсії і вимагав equal_var=False, тут результат
Лівена дозволяє використати equal_var=True.

In [8]:
difference = np.array([
    3, 3, 4, 5, 6, 7,
    7, 7, 6, 5, 4, 3
])

evening = dnipro - difference

print("Ранок:")
print(dnipro)

print("\nВечір:")
print(evening)

Ранок:
[82 79 74 67 61 59 57 59 65 72 78 83]

Вечір:
[79 76 70 62 55 52 50 52 59 67 74 80]


In [9]:
t_rel, p_rel = stats.ttest_rel(
    dnipro,
    evening
)

print("Середнє вранці:", dnipro.mean())
print("Середнє ввечері:", evening.mean())
print("Середня різниця:", (dnipro - evening).mean())

print("t:", t_rel)
print("p-value:", p_rel)

Середнє вранці: 69.66666666666667
Середнє ввечері: 64.66666666666667
Середня різниця: 5.0
t: 10.85620296683619
p-value: 3.234104031121751e-07


Парний t-критерій дав t ≈ 10.86 та
p ≈ 3.23 × 10⁻⁷.

Оскільки p < 0.05, нульову гіпотезу про відсутність
середньої різниці відхиляємо.

Отже, між ранковою та вечірньою вологістю Дніпра
існує статистично значуща різниця.

Середня ранкова вологість становить приблизно 69.67%,
вечірня — 64.67%, тобто середня різниця дорівнює
5 процентним пунктам.

Тут використовується саме парний t-критерій, оскільки
кожне ранкове значення має відповідне вечірнє значення
того самого місяця.

Обчислення різниці всередині кожної пари усуває частину
варіативності, пов'язану з відмінностями між місяцями.
Тому розглядається саме зміна "ранок — вечір", а не дві
незалежні групи.

Контрольні питання

1. Чому для перевірки середнього однієї вибірки
використовується t-, а не z-розподіл?

У задачі справжнє стандартне відхилення генеральної
сукупності σ невідоме.

Ми оцінюємо його за вибіркою за допомогою s.
Через цю додаткову невизначеність використовується
t-розподіл.

При збільшенні розміру вибірки t-розподіл поступово
наближається до стандартного нормального розподілу.


2. Чому неправильний вибір equal_var може дати
неправильний p-value?

Стандартний t-критерій з equal_var=True використовує
припущення, що дисперсії двох генеральних сукупностей
однакові, та об'єднану оцінку дисперсії.

Якщо насправді дисперсії суттєво відрізняються, ця
оцінка та стандартна похибка можуть бути неправильними.
Через це змінюються t-статистика, число ступенів свободи
та p-value.

Тому спочатку використовується критерій Лівена.
Якщо p < 0.05, застосовуємо Welch t-test
(equal_var=False). Якщо p >= 0.05, використовуємо
equal_var=True.


3. Чому парний критерій може бути потужнішим?

У парних даних кожне спостереження однієї вибірки
пов'язане з конкретним спостереженням другої.

Наприклад, ранкова і вечірня вологість виміряні
в одному й тому самому місяці.

При обчисленні різниці всередині кожної пари усувається
частина варіативності між місяцями. Завдяки цьому
залишається менше стороннього розкиду і легше виявити
систематичну різницю між ранком та вечором.

Незалежний критерій ігнорував би цей зв'язок.


4. Чи узгоджується критерій Лівена з порівнянням std?

Так.

У Завданні 3:
Дніпро std ≈ 9.55,
Одеса std ≈ 5.62.

Різниця помітна навіть при простому порівнянні, а
критерій Лівена дав p ≈ 0.016 < 0.05 і підтвердив
статистично значущу різницю дисперсій.

У Завданні 4:
Дніпро std ≈ 9.55,
Київ std ≈ 8.95.

Значення значно ближчі. Критерій Лівена дав
p ≈ 0.785 > 0.05, тому статистично значущої різниці
дисперсій не виявлено.

Отже, формальний результат критерію Лівена узгоджується
з тим, що можна попередньо побачити за стандартними
відхиленнями.